### PR TITLE
CACTUS-459: fix auto-scrolling behavior of MenuBar

### PR DESCRIPTION
**Pull request by wilysword**
https://repayonline.atlassian.net/browse/CACTUS-459

The problem is basically a conflict between browsers' auto-scroll when an element receives focus, and the scroll behavior implemented by the hook. ~To get around this, we have to disable the hook scroll when focus is the result of a click, and then do the scroll after the click.~ (The changes I made to the scrolling algorithm mean scrolling will no longer move the element outside of the click area.) This doesn't 100% fix the problem, though: if you click a submenu item to expand it and the expanded part is completely off-screen, the auto-scroll will still kick in (there's no cross-browser way to prevent that behavior).

This seemed like a good time to refactor the scroll behavior used by both MenuBar and TabList, so I also did that and refined the targeted scrolling algorithm a bit to make it flow a little more naturally. Finally, I changed the tab-focus behavior of the MenuBar to be the same as TabList (changing the tabIndex to -1 on focus, and back to 0 on blur).

### CODE REVIEW

**Review by Dhalton — APPROVED** [review]
LGTM, seems to work pretty well

**Review by mikesoltow — APPROVED** [review]
Looks pretty good to me!  I left a couple of readability comments that you can take or leave.  Also, looks like there are a few merge conflicts.

File: modules/cactus-web/src/MenuBar/MenuBar.tsx [hunk 1]
```diff
@@ -143,80 +125,83 @@ const TextWrapper = styled.div`
...
+const useFocusHandler = (setFocus: FocusSetter) =>
+  React.useCallback(
+    (e: React.FocusEvent<HTMLElement>) => {
+      e.currentTarget.tabIndex = -1
+      if (e.currentTarget === e.target) {
+        setFocus(0) // Set focus on the first menuitem.
+      }
+    },
+    [setFocus]
+  )
```
[the remark this answers is not in the document]
Comment (wilysword) [reply modules/cactus-web/src/MenuBar/MenuBar.tsx]: 👍

File: modules/cactus-web/src/helpers/scroll.ts [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+}
+
+const DEFAULT_SCROLL: Scroll = { showScroll: false, offset: 0, currentIndex: 0 }
+
+// Takes in a list element and returns three things:
+// 1. The outermost wrapper; either the list itself if it's a regular scroll element,
+//    or the container of the list + buttons if it's `overflow: hidden`.
+// 2. The width of one button, or 0 (zero) if it's a regular scroll element.
+// 3. The list of items that can be scrolled to.
+export type GetScrollInfo = (e: HTMLElement) => [HTMLElement, number, HTMLElement[]]
```
[the remark this answers is not in the document]
Comment (wilysword) [reply modules/cactus-web/src/helpers/scroll.ts]: 👍

**Pull request merged**